### DISCUSSION TITLE
Using ui.teleport on Ag grid

### DISCUSSION THREAD

**Discussion by fmcdevelopment** [zauberzeug/nicegui#3665] [Q&A] [answered]
### Question

First of all, Thanks for the relsease of version 2 of NiceGui. It has been awesome🔥!

For  my question, currently, I am trying to change the default floating filter of Ag grid to ui.input which I want to use it later as a custom filter for Ag Grid (using Ag [GridsetFilter](https://www.ag-grid.com/javascript-data-grid/filter-api/#reference--setFilterModel) API). Using ui.teleport, I come up with the code below:

```python
from nicegui import ui

aggrid_table = ui.aggrid({
    'columnDefs': [
        {'headerName': 'Name', 'field': 'name', 'filter': 'agTextColumnFilter', 'floatingFilter': False},
        {'headerName': 'Age', 'field': 'age', 'filter': 'agNumberColumnFilter', 'floatingFilter': True},
    ],
    'rowData': [
        {'name': 'Alice', 'age': 18},
        {'name': 'Bob', 'age': 21},
        {'name': 'Carol', 'age': 42},
    ],
}).classes('max-h-40')
with ui.teleport(f'#c{aggrid_table.id} > div > div.ag-root-wrapper-body.ag-layout-normal.ag-focus-managed > div.ag-root.ag-unselectable.ag-layout-normal > div.ag-header.ag-focus-managed.ag-pivot-off.ag-header-allow-overflow > div.ag-header-viewport > div > div.ag-header-row.ag-header-row-column-filter > div:nth-child(1)'):
    ui.input()

ui.run()
```
But the ui.input is not shown at all. I am new to ui.teleport so can someone point me to the right direction?

The reason behind this usecase is that I want to use Ag Grid [Multifilter](https://www.ag-grid.com/javascript-data-grid/filter-multi/) which is for an enterprise version only.

**Comment by falkoschindler** [comment]
Hi @fmcdevelopment, it is right there:

![Screenshot 2024-09-02 at 16 27 47](https://github.com/user-attachments/assets/289e74aa-f6a6-40c3-ac06-8e298cebb364)

**Comment by fmcdevelopment** [comment]
Oh, when I wrap the code inside page decorator, the ui.teleport seems not working on my machine. Can you take a look at the code below:

```python
from nicegui import ui

@ui.page('/')
async def page_layout():
    aggrid_table = ui.aggrid({
        'columnDefs': [
            {'headerName': 'Name', 'field': 'name', 'filter': 'agTextColumnFilter', 'floatingFilter': False},
            {'headerName': 'Age', 'field': 'age', 'filter': 'agNumberColumnFilter', 'floatingFilter': True},
        ],
        'rowData': [
            {'name': 'Alice', 'age': 18},
            {'name': 'Bob', 'age': 21},
            {'name': 'Carol', 'age': 42},
        ],
    }).classes('max-h-40')
    with ui.teleport(f'#c{aggrid_table.id} > div > div.ag-root-wrapper-body.ag-layout-normal.ag-focus-managed > div.ag-root.ag-unselectable.ag-layout-normal > div.ag-header.ag-focus-managed.ag-pivot-off.ag-header-allow-overflow > div.ag-header-viewport > div > div.ag-header-row.ag-header-row-column-filter > div:nth-child(1)'):
        ui.button('Here it is', on_click=lambda: ui.notify('Hi')).classes('w-full')

ui.run()
```

The resuld should be like this:
<img width="483" alt="Screenshot 2024-09-03 at 2 33 48" src="https://github.com/user-attachments/assets/e3727335-6095-4f74-860b-72c7c5a25e54">

Is it the same on your machine?
I am using M1 Mac Air with the latest NiceGui.

**Reply by falkoschindler** [reply, marked as answer]
That's strange. I can't say for sure why `ui.teleport` doesn't work in this case, but I guess it's some timing issue with how the AG Grid is created on the client.

It looks like you can postpone the teleport element a bit by awaiting the client connection first. This way the button is visible:
```py
    await ui.context.client.connected()
    with ui.teleport(...): ...
```

<img width="485" alt="Screenshot 2024-09-03 at 07 34 55" src="https://github.com/user-attachments/assets/ade0282d-d8ef-4ecf-981d-44031d0d990d">

**Reply by fmcdevelopment** [reply]
Thanks. I also realize that when not using `.classes('max-h-40')` for Ag Grid, the button is working properly from the start:

```python
from nicegui import ui

@ui.page('/')
async def page_layout():
    aggrid_table = ui.aggrid({
        'columnDefs': [
            {'headerName': 'Name', 'field': 'name', 'filter': 'agTextColumnFilter', 'floatingFilter': False},
            {'headerName': 'Age', 'field': 'age', 'filter': 'agNumberColumnFilter', 'floatingFilter': True},
        ],
        'rowData': [
            {'name': 'Alice', 'age': 18},
            {'name': 'Bob', 'age': 21},
            {'name': 'Carol', 'age': 42},
        ],
    })
    with ui.teleport(f'#c{aggrid_table.id} > div > div.ag-root-wrapper-body.ag-layout-normal.ag-focus-managed > div.ag-root.ag-unselectable.ag-layout-normal > div.ag-header.ag-focus-managed.ag-pivot-off.ag-header-allow-overflow > div.ag-header-viewport > div > div.ag-header-row.ag-header-row-column-filter > div:nth-child(1)'):
        ui.button('Here it is', on_click=lambda: ui.notify('Hi')).classes('w-full')

ui.run()
```
But then I can not change the heigth of the Ag Grid. I am quite intrigue by this but I do not have a slightest clue what changes.

as for your suggesstion, I am going to use `await ui.context.client.connected()` but is it possible to use it in two different place inside the same page decorator?
I already am using `await ui.context.client.connected()` for waiting io_bound process as suggested [here](https://github.com/zauberzeug/nicegui/discussions/3477)

**Reply by falkoschindler** [reply]
That's very weird...
No, calling `ui.context.client.connected()` multiple times won't work. The second call will probably run into a timeout.
As a workaround, adding some custom CSS to change the height of AG Grids seems to work:
```py
ui.add_css('.nicegui-aggrid { height: 500px }')
```
Or to be more specific, you can use a custom class:
```py
ui.add_css('.my-aggrid { height: 500px }')
ui.aggrid({...}).classes('my-aggrid')
```
